File: app/src/test/java/tv/blofy/player/core/diagnostics/DiagnosticsSanitizerTest.kt
package tv.blofy.player.core.diagnostics

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class DiagnosticsSanitizerTest {
    @Test
    fun redactsXtreamUserInfoPathQueryAndFragment() {
        val sanitized = DiagnosticsSanitizer.sanitizeUrl(
            "https://authority-user:authority-pass@IPTV.Example:8443/live/alice/p%40ssword/12345.ts" +
                "?username=query-user&password=query-pass&token=query-token#fragment-token"
        )

        assertEquals("https://redacted.invalid/live/***.ts", sanitized)
        listOf(
            "iptv.example", "authority-user", "authority-pass", "alice", "p%40ssword", "query-user",
            "query-pass", "query-token"
        )
            .forEach { secret -> assertFalse(sanitized.contains(secret, ignoreCase = true)) }
    }

    @Test
    fun redactsM3uQueryAndUnknownPaths() {
        assertEquals(
            "https://redacted.invalid/***.php",
            DiagnosticsSanitizer.sanitizeUrl(
                "https://provider.example/get.php?username=alice&password=secret&type=m3u_plus&token=abc"
            )
        )
        assertEquals("<redacted-url>", DiagnosticsSanitizer.sanitizeUrl("not a valid URL?token=abc"))
    }

    @Test
    fun removesEmbeddedUrlsCredentialsAndTokensFromErrors() {
        val sanitized = DiagnosticsSanitizer.sanitizeMessage(
            "Failed https://url-user:url-pass@iptv.example/series/alice/secret/42.mkv?token=url-token " +
                "username=plain-user password: plain-pass token='plain-token' Authorization: Bearer eyJ.secret.sig " +
                "JSON={\"username\":\"json-user\",\"password\":\"json-pass\",\"access_token\":\"json-token\"}"
        ).orEmpty()

        assertTrue(sanitized.contains("<redacted-url>"))
        listOf(
            "iptv.example", "url-user", "url-pass", "alice", "secret", "url-token", "plain-user", "plain-pass",
            "plain-token", "eyJ.secret.sig", "json-user", "json-pass", "json-token"
        ).forEach { secret -> assertFalse("Leaked $secret in: $sanitized", sanitized.contains(secret)) }
        assertFalse(sanitized.contains("http://"))
        assertFalse(sanitized.contains("https://"))
    }

    @Test
    fun removesSchemeLessUrlsAndUriValuesFromErrors() {
        val sanitized = DiagnosticsSanitizer.sanitizeMessage(
            "source=provider.example/live/alice/password/99.ts uri=https%3A%2F%2Fprovider.example%2Fget.php%3Ftoken%3Dabc"
        ).orEmpty()

        assertFalse(sanitized.contains("provider.example"))
        assertFalse(sanitized.contains("alice"))
        assertFalse(sanitized.contains("password/99"))
        assertFalse(sanitized.contains("token%3Dabc"))
    }

    @Test
    fun removesBareHostsIpAddressesBasicAuthAndRelativePaths() {
        val sanitized = DiagnosticsSanitizer.sanitizeMessage(
            "Unable to resolve provider.example.com; ETIMEDOUT 203.0.113.42:8080 " +
                "Authorization: Basic YWxpY2U6c2VjcmV0 uri=/live/path-user/path-pass/1.ts"
        ).orEmpty()

        listOf("provider.example.com", "203.0.113.42", "YWxpY2U6c2VjcmV0", "path-user", "path-pass")
            .forEach { secret -> assertFalse("Leaked $secret in: $sanitized", sanitized.contains(secret)) }
    }

    @Test
    fun pseudonymizesProviderAndAllowlistsMetadata() {
        val provider = DiagnosticsSanitizer.pseudonymizeProviderKey(
            "https://user:pass@provider.example/live/user/pass/1.ts?token=secret"
        )
        assertTrue(provider.matches(Regex("^provider-[a-f0-9]{16}$")))
        assertEquals(provider, DiagnosticsSanitizer.pseudonymizeProviderKey(provider))
        assertFalse(provider.contains("secret"))

        assertEquals("live", DiagnosticsSanitizer.sanitizeContentKind("LIVE"))
        assertEquals("unknown", DiagnosticsSanitizer.sanitizeContentKind("password=secret"))
        assertEquals(
            "ERROR_CODE_IO_NETWORK_CONNECTION_FAILED",
            DiagnosticsSanitizer.sanitizeErrorCode("ERROR_CODE_IO_NETWORK_CONNECTION_FAILED")
        )
        assertEquals(null, DiagnosticsSanitizer.sanitizeErrorCode("token=secret"))
        assertEquals("2.0.0-rc01", DiagnosticsSanitizer.sanitizeAppVersion("2.0.0-rc01"))
        assertEquals(null, DiagnosticsSanitizer.sanitizeAppVersion("url=https://provider.example"))
    }

    @Test
    fun enforcesOutputLimitsAfterRedaction() {
        assertEquals(24, DiagnosticsSanitizer.sanitizeUrl("https://example.com/live/u/p/1.ts", 24).length)
        assertEquals(32, DiagnosticsSanitizer.sanitizeMessage("failure " + "x".repeat(200), 32)?.length)
    }
}

File: services/activation/test/diagnostics-sanitizer.test.mjs
import assert from 'node:assert/strict';
import test from 'node:test';
import {
  pseudonymizeDiagnosticProviderKey,
  safeErrorSummary,
  sanitizeDiagnosticAppVersion,
  sanitizeDiagnosticContentKind,
  sanitizeDiagnosticErrorCode,
  sanitizeDiagnosticMessage,
  sanitizeDiagnosticUrl
} from '../src/diagnostics-sanitizer.mjs';

test('redacts URL user-info, Xtream path credentials, query and fragment', () => {
  const sanitized = sanitizeDiagnosticUrl(
    'https://authority-user:authority-pass@IPTV.Example:8443/live/alice/p%40ssword/12345.ts' +
      '?username=query-user&password=query-pass&token=query-token#fragment-token'
  );

  assert.equal(sanitized, 'https://redacted.invalid/live/***.ts');
  for (const secret of [
    'iptv.example', 'authority-user', 'authority-pass', 'alice', 'p%40ssword', 'query-user', 'query-pass',
    'query-token'
  ]) {
    assert.equal(sanitized.toLowerCase().includes(secret.toLowerCase()), false, `leaked ${secret}`);
  }
});

test('redacts M3U query credentials and malformed input', () => {
  assert.equal(
    sanitizeDiagnosticUrl('https://provider.example/get.php?username=alice&password=secret&type=m3u_plus&token=abc'),
    'https://redacted.invalid/***.php'
  );
  assert.equal(sanitizeDiagnosticUrl('not a valid URL?token=abc'), '<redacted-url>');
});

test('removes embedded URLs, username, password and token values from errors', () => {
  const sanitized = sanitizeDiagnosticMessage(
    'Failed https://url-user:url-pass@iptv.example/series/alice/secret/42.mkv?token=url-token ' +
      'username=plain-user password: plain-pass token=\'plain-token\' Authorization: Bearer eyJ.secret.sig ' +
      'JSON={"username":"json-user","password":"json-pass","access_token":"json-token"}'
  );

  assert.match(sanitized, /<redacted-url>/);
  for (const secret of [
    'iptv.example', 'url-user', 'url-pass', 'alice', 'secret', 'url-token', 'plain-user', 'plain-pass',
    'plain-token', 'eyJ.secret.sig', 'json-user', 'json-pass', 'json-token'
  ]) {
    assert.equal(sanitized.includes(secret), false, `leaked ${secret} in: ${sanitized}`);
  }
  assert.equal(sanitized.includes('http://'), false);
  assert.equal(sanitized.includes('https://'), false);
});

test('removes scheme-less URLs and encoded URI values from errors', () => {
  const sanitized = sanitizeDiagnosticMessage(
    'source=provider.example/live/alice/password/99.ts ' +
      'uri=https%3A%2F%2Fprovider.example%2Fget.php%3Ftoken%3Dabc'
  );

  for (const secret of ['provider.example', 'alice', 'password/99', 'token%3Dabc']) {
    assert.equal(sanitized.includes(secret), false, `leaked ${secret} in: ${sanitized}`);
  }
});

test('removes bare hosts, IP addresses and Basic authorization credentials', () => {
  const sanitized = sanitizeDiagnosticMessage(
    'Unable to resolve host "provider.example.com"; ETIMEDOUT 203.0.113.42:8080 ' +
      'Authorization: Basic YWxpY2U6c2VjcmV0'
  );

  for (const secret of ['provider.example.com', '203.0.113.42', 'YWxpY2U6c2VjcmV0']) {
    assert.equal(sanitized.includes(secret), false, `leaked ${secret} in: ${sanitized}`);
  }
});

test('removes credential-bearing relative stream paths', () => {
  const sanitized = sanitizeDiagnosticMessage(
    'DataSpec uri=/live/relative-user/relative-pass/123.ts and (/series/other-user/other-pass/9.mkv)'
  );

  for (const secret of ['relative-user', 'relative-pass', 'other-user', 'other-pass']) {
    assert.equal(sanitized.includes(secret), false, `leaked ${secret} in: ${sanitized}`);
  }
  assert.match(sanitized, /<redacted-url>/);
});

test('pseudonymizes provider keys and allowlists all remaining diagnostics metadata', () => {
  const provider = pseudonymizeDiagnosticProviderKey(
    'https://user:pass@provider.example/live/user/pass/1.ts?token=secret'
  );
  assert.match(provider, /^provider-[a-f0-9]{16}$/);
  assert.equal(pseudonymizeDiagnosticProviderKey(provider), provider);
  assert.equal(provider.includes('secret'), false);

  assert.equal(sanitizeDiagnosticContentKind('LIVE'), 'live');
  assert.equal(sanitizeDiagnosticContentKind('password=secret'), 'unknown');
  assert.equal(sanitizeDiagnosticErrorCode('ERROR_CODE_IO_NETWORK_CONNECTION_FAILED'), 'ERROR_CODE_IO_NETWORK_CONNECTION_FAILED');
  assert.equal(sanitizeDiagnosticErrorCode('token=secret'), null);
  assert.equal(sanitizeDiagnosticAppVersion('2.0.0-rc01'), '2.0.0-rc01');
  assert.equal(sanitizeDiagnosticAppVersion('url=https://provider.example'), null);
});

test('limits only sanitized output', () => {
  assert.equal(sanitizeDiagnosticUrl('https://example.com/live/u/p/1.ts', 24).length, 24);
  assert.equal(sanitizeDiagnosticMessage(`failure ${'x'.repeat(200)}`, 32).length, 32);
});

test('server error summaries never serialize messages, stacks, causes, URLs or credentials', () => {
  const error = new Error(
    'password=SUPER-SECRET token=TOP-TOKEN url=https://user:pass@provider.example/live/user/pass/1.ts'
  );
  error.stack = `STACK WITH SUPER-SECRET ${error.stack}`;
  error.cause = { connectionString: 'postgresql://db-user:db-pass@database.example/db' };
  error.request = { body: { username: 'playlist-user', password: 'playlist-pass' } };
  error.code = 'ECONNREFUSED';

  assert.equal(safeErrorSummary(error), 'Error (ECONNREFUSED)');

  error.code = 'ECONNREFUSED token=LEAK';
  assert.equal(safeErrorSummary(error), 'Error');
});
